From 56c7d81351b68e33e6424a7c548d4106f327bba4 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Tue, 28 Nov 2023 18:26:57 +0200
Subject: [PATCH] [FSW-334] fs1: enable pci drain

Pci drain is enabled and the drain stamp that register the address
of the timed-out trasaction is enabled as well.

Change-Id: I8637a7412d55d0328db981683efec9c3669b178f
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 01e078e..ec86362 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -14,6 +14,8 @@
 #include <psoc_cfg_misc_regs.h>
 #include <int_agg_psoc_i2c_slv_comb_regs.h>
 #include <pll_ctrl_regs.h>
+#include <cxl_ep_cfg_pcie_wrap_regs.h>
+#include <cxl_ep_cfg_pcie_wrap_masks.h>
 #include <fs1_blocks_linux_driver.h>
 
 #include <baremetal.h>
@@ -143,6 +145,37 @@
 	pll.div_factor_offset = mmPLL_CTRL_DIV_FACTOR_0;
 }
 
+
+#define HBW_DRAIN_TIMEOUT	0x1000
+#define LBW_DRAIN_TIMEOUT	0xF000
+
+void ppboot_pcie_drain_init(void)
+{
+	u32 val;
+
+	regmap_writel(HBW_DRAIN_TIMEOUT, mmIO_0_CFG_HALF_1_CXL_EP_PCIE_WRAP_BASE +
+					 mmCXL_EP_CFG_PCIE_WRAP_HBW_DRAIN_TIMEOUT);
+	regmap_writel(CXL_EP_CFG_PCIE_WRAP_HBW_RSP_ERR_DRAIN_STAMP_EN_M,
+		      mmIO_0_CFG_HALF_1_CXL_EP_PCIE_WRAP_BASE +
+		      mmCXL_EP_CFG_PCIE_WRAP_HBW_RSP_ERR_DRAIN_STAMP);
+	val = regmap_readl(mmIO_0_CFG_HALF_1_CXL_EP_PCIE_WRAP_BASE +
+			   mmCXL_EP_CFG_PCIE_WRAP_HBW_DRAIN_CFG);
+	val |= CXL_EP_CFG_PCIE_WRAP_HBW_DRAIN_CFG_EN_M;
+	regmap_writel(val, mmIO_0_CFG_HALF_1_CXL_EP_PCIE_WRAP_BASE +
+			   mmCXL_EP_CFG_PCIE_WRAP_HBW_DRAIN_CFG);
+
+	regmap_writel(LBW_DRAIN_TIMEOUT, mmIO_0_CFG_HALF_1_CXL_EP_PCIE_WRAP_BASE +
+					 mmCXL_EP_CFG_PCIE_WRAP_LBW_DRAIN_TIMEOUT);
+	regmap_writel(CXL_EP_CFG_PCIE_WRAP_LBW_RSP_ERR_DRAIN_STAMP_EN_M,
+		      mmIO_0_CFG_HALF_1_CXL_EP_PCIE_WRAP_BASE +
+		      mmCXL_EP_CFG_PCIE_WRAP_LBW_RSP_ERR_DRAIN_STAMP);
+	val = regmap_readl(mmIO_0_CFG_HALF_1_CXL_EP_PCIE_WRAP_BASE +
+			   mmCXL_EP_CFG_PCIE_WRAP_LBW_DRAIN_CFG);
+	val |= CXL_EP_CFG_PCIE_WRAP_LBW_DRAIN_CFG_EN_M;
+	regmap_writel(val, mmIO_0_CFG_HALF_1_CXL_EP_PCIE_WRAP_BASE +
+			   mmCXL_EP_CFG_PCIE_WRAP_LBW_DRAIN_CFG);
+}
+
 void preboot_sync_btl(void)
 {
 	/* not relevant in case of secure boot flow */
